### ISSUE TITLE
Test item GetAssertionEmptyUserIdTest

### ISSUE THREAD

**Issue by nuno0529** [google/CTAP2-test-tool#90]
{
      "description": "Tests if empty user IDs are omitted in the response.",
      "error_message": "Cannot make credential with an empty user ID.",
      "id": "get_assertion_empty_user_id",
      "observations": [
        "A prompt was expected, but not performed. Sometimes it is just not recognized if performed too fast.",
        "The failing error code is `CTAP2_ERR_PROCESSING`."
      ],
      "result": "fail",
      "tags": []
    },

https://github.com/fido-alliance/fido-2-specs/pull/963
Although fido2 spec says empty account identifier is valid but can't be returned in get(), this case makes authenticator need to save some useless data...

**Comment by kaczmarczyck** [comment]
Thanks for opening the specification issue. We can follow up here after the discussion on the specification.

**Comment by nuno0529** [comment]
Another question, shouldn't this test case (i.e. GetAssertionEmptyUserIdTest ) use rk=true in makeCred command or normal authenticator (w/o force to create rk) won't return any user data in getAssert response.

**Comment by kaczmarczyck** [comment]
Good point, `rk=true` would more likely trigger bad behaviour. Also, I'll follow up on the specification isse.

**Comment by nuno0529** [comment]
For this test case after related information from ctap2/webauth, I think the exepected reuslts should be 
1. (unchanged) Authenticator MUST accept makCred command with empty user ID with rk.
1. (changed) Authenticator MUST return resident credential with empty user ID in geAssert response.

**Comment by kaczmarczyck** [comment]
I can only change this test to the changed behaviour after confirming it doesn't break Windows support. At the moment, this tool is used to test firmware releases, so I have to test what actually works, even if it contradicts the specification.
@nuno0529 Have you tested your proposed behaviour on Windows? We had problems with OpenSK, i.e.
[https://github.com/google/OpenSK/issues/28](https://github.com/google/OpenSK/issues/28)

**Comment by kaczmarczyck** [comment]
WebAuthn adapted just now:
[https://github.com/w3c/webauthn/pull/1537](https://github.com/w3c/webauthn/pull/1537)

**Comment by nuno0529** [comment]
I didn't test so many fido2 test servers which are for test purpose only. At least I didn't encounter this issue on mainstream websites  and main test server webauthntest.azurewebsites.net I used. So I'm still thinking if authenticators do need to apply this kind of workaround behavior for this issue (i.e. not return user info with empty user.id).

**Comment by kaczmarczyck** [comment]
May I ask you to test on [https://webauthn.me/](https://webauthn.me/) on a Windows 10, with UV being turned on?

**Comment by nuno0529** [comment]
I don't see any problem when testing on https://webauthn.me/ and this doesn't use empty user.id in create().

**Comment by kaczmarczyck** [comment]
I wanted to get an ATKey.pro to get to the bottom of this, but Amazon doesn't ship it to Europe it seems. I assume this is one of the device I could use to reproduce? Any other way to get one of them?

**Comment by nuno0529** [comment]
We can send you one to Europe, please contact jenny.hsu@authentrend.com with your address and I have notify her this request from @kaczmarczyck.

But if the issue you want to reproduce is "Cannot make credential with an empty user ID." I can also reproduce this behavior with Yubikeys and the log here also show this error.
https://github.com/google/CTAP2-test-tool/blob/master/results/Security%20Key%20by%20Yubico_.json

**Comment by kaczmarczyck** [comment]
Thanks, email is out! Just making sure, and for documentation porpuses: For reproducing on [https://webauthn.me/](https://webauthn.me/), you have to be on Windows 10 and your authenticator needs to be set up with UV.

**Comment by nuno0529** [comment]
> For reproducing on https://webauthn.me/, you have to be on Windows 10 and your authenticator needs to be set up with UV.

I test on Windows 10 20H2(19042.685) with MS Edge 87.0.664.60, I use my authenticator with built-in UV and I don't see any problem or and special parts related to this issue. Do you mean under https://webauthn.me/debugger ? But the user.id here still can't be set to empty or any else.

**Comment by kaczmarczyck** [comment]
I'll debug locally (after Christmas) and come back to you. Thanks so far for the detailed information!

**Comment by kaczmarczyck** [comment]
Update: We still have the same behaviour for OpenSK on Windows for [webauthn.me](webauthn.me) with UV. `GetAssertion` fails if we remove [this check](https://github.com/google/OpenSK/blob/b0c1b73897816edb757a163943f7b36a55f76a19/src/ctap/mod.rs#L657).

We can leave this issue open for tracking, but I think it currently a good compromise between specification and working in the real world.
